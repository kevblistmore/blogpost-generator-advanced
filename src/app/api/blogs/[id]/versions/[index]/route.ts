// src/app/api/blogs/[id]/versions/[index]/route.ts
import { connectDB } from '../../../../../lib/db';
import { Blog } from '../../../../../lib/db';
import { NextResponse } from 'next/server';

export async function DELETE(
  req: Request,
  { params }: { params: { id: string; index: string } }
) {
  try {
    await connectDB();
    const blog = await Blog.findById(params.id);
    
    if (!blog) {
      return NextResponse.json({ error: 'Blog not found' }, { status: 404 });
    }

    const index = parseInt(params.index);
    if (isNaN(index) || index >= blog.versions.length || index < 0) {
      return NextResponse.json({ error: 'Invalid version index' }, { status: 400 });
    }

    // Prevent deleting original version
    if (index === 0) {
      return NextResponse.json(
        { error: 'Cannot delete original version' },
        { status: 400 }
      );
    }

    blog.versions.splice(index, 1);
    blog.currentVersion = Math.min(blog.currentVersion, blog.versions.length - 1);
    await blog.save();

    return NextResponse.json(blog.toObject());
  } catch (error) {
    return NextResponse.json(
      { error: 'Failed to delete version' },
      { status: 500 }
    );
  }
}