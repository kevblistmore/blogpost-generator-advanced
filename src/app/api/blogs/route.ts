//src/app/api/blogs/route.ts
import { connectDB } from '../../lib/db';
import { Blog } from '../../lib/db';
import { NextResponse } from 'next/server';

export async function GET() {
  try {
    await connectDB();
    const blogs = await Blog.find().sort({ createdAt: -1 }).lean();
    return NextResponse.json(blogs);
  } catch (error) {
    return NextResponse.json(
      { error: 'Failed to fetch blogs' },
      { status: 500 }
    );
  }
}

export async function POST(req: Request) {
  try {
    const { content, topic } = await req.json();
    await connectDB();
    
    const blog = new Blog({
      content,
      topic,
      images: []
    });

    await blog.save();
    return NextResponse.json(blog.toObject());
  } catch (error) {
    return NextResponse.json(
      { error: 'Failed to save blog' },
      { status: 500 }
    );
  }
}

export async function DELETE(req: Request) {
  try {
    const { id } = await req.json();
    await connectDB();
    await Blog.findByIdAndDelete(id);
    return NextResponse.json({ success: true });
  } catch (error) {
    return NextResponse.json(
      { error: 'Failed to delete blog' },
      { status: 500 }
    );
  }
}