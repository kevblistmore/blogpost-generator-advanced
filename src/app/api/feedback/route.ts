// app/api/feedback/route.ts
import { NextRequest, NextResponse } from 'next/server';

export async function POST(request: NextRequest) {
  try {
    const { content, rating, timestamp } = await request.json();
    // In production: store feedback in a DB
    console.log('User Feedback:', { content, rating, timestamp });
    return NextResponse.json({ success: true });
  } catch (error) {
    console.error('Feedback submission failed:', error);
    return NextResponse.json(
      { error: 'Failed to submit feedback' },
      { status: 500 }
    );
  }
}
