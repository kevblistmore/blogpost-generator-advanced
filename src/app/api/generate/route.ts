// app/api/generate/route.ts
import { NextResponse } from 'next/server';
import { Configuration, OpenAIApi } from 'openai';

const configuration = new Configuration({
  apiKey: process.env.OPENAI_API_KEY,
});
const openai = new OpenAIApi(configuration);

export async function POST(request: Request) {
  try {
    const body = await request.text();
    if (!body) {
      throw new Error('Request body is empty');
    }
    const { topic, title } = JSON.parse(body);

    // More explicit markdown prompt:
    const promptText = `
      Write a comprehensive blog post about "${title || topic}" using markdown formatting only.
      **Do not** wrap the entire text in triple backticks.
      Use headings, paragraphs, and lists in Markdown.
      Return the content directly without any code fences.
      `;
    
    const response = await openai.createChatCompletion({
      model: 'gpt-4',
      messages: [{ role: 'user', content: promptText }],
      temperature: 0.7,
    });
    
    const content = response.data.choices[0]?.message?.content || '';
    //console.log('Generated content:', content);
    return NextResponse.json({ content });
    
  } catch (error) {
    console.error('Generation error:', error);
    return NextResponse.json(
      { error: 'Failed to generate content' },
      { status: 500 }
    );
  }
}

// // app/api/generate/route.ts
// import { NextResponse } from 'next/server';
// import { Configuration, OpenAIApi } from 'openai';

// const configuration = new Configuration({
//   apiKey: process.env.OPENAI_API_KEY,
// });
// const openai = new OpenAIApi(configuration);

// export async function POST(request: Request) {
//   try {
//     const { topic, title,previousContent } = await request.json();
    
//     // Only include previous content if it exists
//     const extraPrompt = previousContent ? previousContent.slice(0, 500) : '';

//     // More explicit markdown prompt:
//     const promptText = `
//       Write a comprehensive blog post about "${title || topic}" using markdown formatting only.
//       **Do not** wrap the entire text in triple backticks.
//       Use headings, paragraphs, and lists in Markdown.
//       Return the content directly without any code fences.
//       ${extraPrompt}...
//     `;
    
//     const response = await openai.createChatCompletion({
//       model: 'gpt-4o',
//       messages: [{ role: 'user', content: promptText }],
//       temperature: 0.7,
//     });
    
//     const content = response.data.choices[0]?.message?.content || '';
//     //console.log('Generated content:', content);
//     return NextResponse.json({ content });
    
//   } catch (error) {
//     console.error('Generation error:', error);
//     return NextResponse.json(
//       { error: 'Failed to generate content' },
//       { status: 500 }
//     );
//   }
// }
