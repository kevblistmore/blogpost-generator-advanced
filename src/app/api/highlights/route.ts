// src/app/api/highlights/route.ts
import { NextResponse } from "next/server";
import { Configuration, OpenAIApi } from "openai";

const configuration = new Configuration({
  apiKey: process.env.OPENAI_API_KEY,
});
const openai = new OpenAIApi(configuration);

export async function POST(request: Request) {
  try {
    const { content } = await request.json();
    const prompt = `Extract the following from the blog post as a JSON object with keys "keywords", "phrases", and "summary". Each key should have an array of strings.
Blog post:
${content}`;

    const response = await openai.createChatCompletion({
      model: "gpt-4o",
      messages: [{ role: "user", content: prompt }],
      temperature: 0.7,
    });

    const rawHighlights = response.data.choices[0]?.message?.content || "[]";
    // Clean up the output if needed
    const cleanedHighlights = rawHighlights
      .replace(/```json/g, "")
      .replace(/```/g, "")
      .trim();

    const highlights = JSON.parse(cleanedHighlights);
    return NextResponse.json({ highlights });
  } catch (error) {
    console.error("Highlights error:", error);
    return NextResponse.json({ highlights: [] }, { status: 500 });
  }
}
