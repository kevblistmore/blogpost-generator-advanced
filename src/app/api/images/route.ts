import { NextResponse } from 'next/server';

export async function GET(req: Request) {
  const { searchParams } = new URL(req.url);
  const topic = searchParams.get('topic') || 'technology';
  
  // Using Pexels API (replace with your API key)
  const API_KEY = process.env.PEXELS_API_KEY!;
  const url = `https://api.pexels.com/v1/search?query=${topic}&per_page=6`;

  try {
    const res = await fetch(url, {
      headers: { Authorization: API_KEY }
    });
    
    if (!res.ok) throw new Error('Image fetch failed');
    
    const data = await res.json();
    const urls = data.photos?.map((photo: { src: { medium: string } }) => photo.src.medium) || [];
    
    return NextResponse.json({ urls });
  } catch (_error) {
    return NextResponse.json(
      { urls: [] },
      { status: 200 }
    );
  }
}