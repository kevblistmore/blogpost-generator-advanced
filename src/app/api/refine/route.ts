// src/app/api/refine/route.ts
import { NextResponse } from "next/server";
import { Configuration, OpenAIApi } from "openai";

const configuration = new Configuration({
  apiKey: process.env.OPENAI_API_KEY,
});
const openai = new OpenAIApi(configuration);

export async function POST(request: Request) {
  try {
    const { content, feedback } = await request.json();
    const prompt = `Here is a blog post:
      ${content}

      User feedback: ${feedback}

      Please provide a refined version of the blog post based on the feedback.`;

    const response = await openai.createChatCompletion({
      model: "gpt-4o",
      messages: [{ role: "user", content: prompt }],
      temperature: 0.7,
    });

    const refinedContent = response.data.choices[0]?.message?.content || "";
    console.log("Refined content:", refinedContent);
    return NextResponse.json({ 
      refinedContent,
      previousContent: content // Include previous content for versioning
    });
  } catch (error) {
    console.error("Refinement error:", error);
    return NextResponse.json({ error: "Failed to refine content" }, { status: 500 });
  }
}
