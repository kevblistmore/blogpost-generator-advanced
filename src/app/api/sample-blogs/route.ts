// app/api/sample-blogs/route.ts
import { NextResponse } from 'next/server';
import sampleBlogs from '../../../../samples/sample-blogs.json';

export async function GET() {
  return NextResponse.json(sampleBlogs);
}
