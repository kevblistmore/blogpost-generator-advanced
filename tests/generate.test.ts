import { NextResponse } from 'next/server';
import { POST as generateHandler } from '../src/app/api/generate/route';

// 1) Mock the entire openai library
jest.mock('openai', () => {
  const mockCreateChatCompletion = jest.fn();

  // Return object that mimics structure of 'openai'
  return {
    // Mocked Config constructor
    Configuration: jest.fn().mockImplementation(() => ({})),

    // Mock OAI client
    OpenAIApi: jest.fn().mockImplementation(() => {
      return {
        createChatCompletion: mockCreateChatCompletion,
      };
    }),

    // For convenience, export mock so we can reference it in our tests
    __mockCreateChatCompletion: mockCreateChatCompletion,
  };
});

// 2) Extract the mock function so we can manipulate or inspect it
const { __mockCreateChatCompletion } = jest.requireMock('openai');

describe('POST /api/generate (Advanced Tests)', () => {
  beforeEach(() => {
    jest.clearAllMocks(); 
  });

  it('should call OpenAI with correct prompt and return the content', async () => {
    // Set up mock result from OpenAI
    __mockCreateChatCompletion.mockResolvedValueOnce({
      data: {
        choices: [
          {
            message: {
              content: 'Mocked AI blog output'
            }
          }
        ]
      }
    });

    // Prepare request with topic + title
    const request = new Request('http://localhost:3000/api/generate', {
      method: 'POST',
      body: JSON.stringify({
        topic: 'Test Topic',
        title: 'My Great Title'
      })
    });

    // Call route
    const response = await generateHandler(request);
    const json = await response.json();

    // Basic checks
    expect(response).toBeInstanceOf(NextResponse);
    expect(json.content).toBe('Mocked AI blog output');

    // 3) Verify the route calls OpenAIApi with certain arguments
    expect(__mockCreateChatCompletion).toHaveBeenCalledTimes(1);
    expect(__mockCreateChatCompletion).toHaveBeenCalledWith({
      model: 'gpt-4',
      messages: [
        {
          role: 'user',
          content: expect.stringContaining('Write a comprehensive blog post')
        }
      ],
      temperature: 0.7,
    });
  });


  it('returns 500 if OpenAI throws an error', async () => {
    // Force an error in the mock
    __mockCreateChatCompletion.mockRejectedValueOnce(new Error('OpenAI failure'));

    const request = new Request('http://localhost:3000/api/generate', {
      method: 'POST',
      body: JSON.stringify({
        topic: 'Bad topic',
        title: ''
      })
    });

    const response = await generateHandler(request);
    const json = await response.json();

    expect(response.status).toBe(500);
    expect(json.error).toBe('Failed to generate content');
  });


  it('returns 500 if request body is invalid (missing JSON)', async () => {
    // No body at all
    const request = new Request('http://localhost:3000/api/generate', {
      method: 'POST',
    });

    // The route's code might throw on request.json() or guard it
    // You may wrap your route with try/catch or check if body is empty first
    const response = await generateHandler(request);
    const json = await response.json();

    // Depending on your route code logic, you might handle it differently,
    // but let's assume it fails
    expect(response.status).toBe(500);
    expect(json.error).toMatch(/Failed to generate content/i);
  });


  it('snapshot test for stable AI output', async () => {
    __mockCreateChatCompletion.mockResolvedValueOnce({
      data: {
        choices: [
          {
            message: {
              content: `
Mocked AI blog output with Markdown
# Title
- bullet A
- bullet B
`
            }
          }
        ]
      }
    });

    const request = new Request('http://localhost:3000/api/generate', {
      method: 'POST',
      body: JSON.stringify({
        topic: 'Snapshot Topic'
      })
    });

    const response = await generateHandler(request);
    const { content } = await response.json();

    // Store a snapshot of the content. If it changes unexpectedly,
    // Jest will report a mismatch
    expect(content).toMatchSnapshot();
  });
});

// import { NextResponse } from 'next/server';
// import { POST as generateHandler } from '../src/app/api/generate/route';

// // Mock the openai library so we don't make real calls:
// jest.mock('openai', () => {
//   return {
//     Configuration: jest.fn().mockImplementation(() => ({})),
//     OpenAIApi: jest.fn().mockImplementation(() => {
//       return {
//         createChatCompletion: jest.fn().mockResolvedValue({
//           data: {
//             choices: [
//               { 
//                 message: { 
//                   content: 'Mocked AI blog content' 
//                 } 
//               }
//             ]
//           }
//         })
//       };
//     })
//   };
// });

// describe('POST /api/generate', () => {
//   it('returns generated blog content from OpenAI', async () => {
//     // A mock Next.js Request object:
//     const request = new Request('http://localhost:3000/api/generate', {
//       method: 'POST',
//       body: JSON.stringify({
//         topic: 'Test Topic',
//         title: 'My Great Title'
//       })
//     });

//     const response = await generateHandler(request);
//     const json = await response.json();

//     // Validate the result
//     expect(response).toBeInstanceOf(NextResponse);
//     expect(json.content).toBe('Mocked AI blog content');
//   });
// });