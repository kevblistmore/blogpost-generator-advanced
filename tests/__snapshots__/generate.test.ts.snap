// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`POST /api/generate (Advanced Tests) snapshot test for stable AI output 1`] = `
"
Mocked AI blog output with Markdown
# Title
- bullet A
- bullet B
"
`;
